syz-manager/manager.go

```diff
@@ -1445,46 +1447,84 @@ func (mgr *Manager) machineChecked(features flatrpc.Feature, enabledSyscalls map
 	mgr.checkDone = true
 	mgr.enabledFeatures = features
 	mgr.targetEnabledSyscalls = enabledSyscalls
+	mgr.firstConnect.Store(time.Now().Unix())
 	statSyscalls := stats.Create("syscalls", "Number of enabled syscalls",
 		stats.Simple, stats.NoGraph, stats.Link("/syscalls"))
 	statSyscalls.Add(len(enabledSyscalls))
 
-	rnd := rand.New(rand.NewSource(time.Now().UnixNano()))
-	fuzzerObj := fuzzer.NewFuzzer(context.Background(), &fuzzer.Config{
-		Corpus:         mgr.corpus,
-		BaseOpts:       opts,
-		Coverage:       mgr.cfg.Cover,
-		FaultInjection: features&flatrpc.FeatureFault != 0,
-		Comparisons:    features&flatrpc.FeatureComparisons != 0,
-		Collide:        true,
-		EnabledCalls:   enabledSyscalls,
-		NoMutateCalls:  mgr.cfg.NoMutateCalls,
-		FetchRawCover:  mgr.cfg.RawCover,
-		Logf: func(level int, msg string, args ...interface{}) {
-			if level != 0 {
-				return
-			}
-			log.Logf(level, msg, args...)
-		},
-		NewInputFilter: func(call string) bool {
-			mgr.mu.Lock()
-			defer mgr.mu.Unlock()
-			return !mgr.saturatedCalls[call]
-		},
-	}, rnd, mgr.target)
-	mgr.fuzzer.Store(fuzzerObj)
+	corpus := mgr.loadCorpus()
+	mgr.phase = phaseLoadedCorpus
 
-	mgr.loadCorpus()
-	mgr.firstConnect.Store(time.Now().Unix())
-	go mgr.corpusMinimization()
-	go mgr.fuzzerLoop(fuzzerObj)
-	if mgr.dash != nil {
-		go mgr.dashboardReporter()
-		if mgr.cfg.Reproduce {
-			go mgr.dashboardReproTasks()
+	if mgr.mode == ModeFuzzing {
+		rnd := rand.New(rand.NewSource(time.Now().UnixNano()))
+		fuzzerObj := fuzzer.NewFuzzer(context.Background(), &fuzzer.Config{
+			Corpus:         mgr.corpus,
+			BaseOpts:       opts,
+			Coverage:       mgr.cfg.Cover,
+			FaultInjection: features&flatrpc.FeatureFault != 0,
+			Comparisons:    features&flatrpc.FeatureComparisons != 0,
+			Collide:        true,
+			EnabledCalls:   enabledSyscalls,
+			NoMutateCalls:  mgr.cfg.NoMutateCalls,
+			FetchRawCover:  mgr.cfg.RawCover,
+			Logf: func(level int, msg string, args ...interface{}) {
+				if level != 0 {
+					return
+				}
+				log.Logf(level, msg, args...)
+			},
+			NewInputFilter: func(call string) bool {
+				mgr.mu.Lock()
+				defer mgr.mu.Unlock()
+				return !mgr.saturatedCalls[call]
+			},
+		}, rnd, mgr.target)
+		fuzzerObj.AddCandidates(corpus)
+		mgr.fuzzer.Store(fuzzerObj)
+
+		go mgr.corpusMinimization()
+		go mgr.fuzzerLoop(fuzzerObj)
+		if mgr.dash != nil {
+			go mgr.dashboardReporter()
+			if mgr.cfg.Reproduce {
+				go mgr.dashboardReproTasks()
+			}
 		}
+		return fuzzerObj
+	} else if mgr.mode == ModeCorpusRun {
+		return mgr.setupCorpusRunner(opts, corpus)
 	}
-	return fuzzerObj
+	panic(fmt.Sprintf("unexpected mode %q", mgr.mode))
+}
+
+func (mgr *Manager) setupCorpusRunner(opts flatrpc.ExecOpts, candidates []fuzzer.Candidate) queue.Source {
+	// Keep `size` elements in the queue.
```

Comment:
A simpler way would be to have a Source that returns a random corpus element in Next.
Is this implementation better? Did you want to ensure that all elements executed the same number of times?

Reply:
I think we do want to make sure that all programs are executed at least once. If we always pick random ones, we are not guaranteed to run all, especially on a corpus with tens of thousands of programs.

You are right that I could have indeed just provided a custom `Next()` implementation. That would be simpler than to track the statuses of individual requests.

Reply:
I've refactored the code to first run all elements of the corpus once and then pick random ones.